docs/internal/compiler-steps.md

```diff
@@ -0,0 +1,119 @@
+
+The stages (and status) of the parser are:
+* (mostly done) Lex/Parse (+ syntax check)
+* (mostly done) Resolve symbols
+* (in progress) Type inference + Generics + Reification
+* (in progress) Reachability analysis & code elision
+* (mostly done) Emit A-normal form + Symbol table
+
+The parser code can be found at `src/parser`.
+
+## MLIR
+
+The MLIR stage takes in the AST and symbol table above and lowers to the following MLIR dialects:
+* Standard: Overall structure (functions, modules, etc) + existing operations
+* LLVM: Structures, memory management, remaining operations
+* Region (new): A dialect that tags objects into regions + creation & destruction semantics.
+
+With the (high-level) IR above, the following MLIR passes are executed:
+* Region and alias analysis
+* Heap to stack transformation
+* Foreign code RPC generation
+
+The final lowering to LLVM IR is done via:
+* Partial lowering from the Region dialect to Std/LLVM dialects
+* Final lowering to LLVM dialect and LLVM IR
+
+The stages (and status) of the MLIR generator are:
+* (in progress) AST lowering to high-level MLIR
+* (not started) Optimisation passes
+* (in progress) Final lowering to LLVM IR
+
+The MLIR code can be found at `src/mlir`.
+
+## LLVM
+
+This stage is mainly about managing the LLVM compiler to produce object code, but it will need creation of new passes and analises.
+
+The Verona LLVM driver will need to:
+* Create a new pass manager infrastructure
+* Create new passes to handle Verona-specific changes
+* Register the existing passes intercalated with our new passes
+* Drive the code through the pipeline
+* Select the correct back-end options to emit appropriate object code
+
+An initial take on which passes we'll need:
+* Clean ups, mem2reg, constant folding and propagation, etc.
+* Early inline pass, at least all functions marked *always_inline* plus obvious ones (arithmetic)
+* Whole program analysis (reachability, multi-versioning)
+* More cleanups and late inlining
+* Code optimisations (inst.combine, loop opts, vectorisation, etc.)
+* Final cleanups
+
+The back-end will need information from the environment and target flags:
+* Target triple defines the back-end to use as well as defaul options
+* Remaining target flags overwrite some defaults
+* Environment can fill other gaps when needed
+
+The stages (and status) of the LLVM driver are:
+* (in progress) Main driver and pass ordering
+* (not started) New passes
+* (not started) Target options
+* (not started) Object code generation
+
+The LLVM code can be found at `src/mlir`.
+
+## Linker
+
+The final output of the compiler is a single object file containing the whole Verona meta-module and all its compile-time dependencies.
+
+In addition to those, Verona will have the following components to link:
+* The Verona runtime, declared at compile time but not implemented
```

Comment:
This should be linked as IR, not a .o file, so we can get inlining of the runtime into the rest of the system.

Reply:
Both runtime and builtins are probably going to be large pieces, and if they're kept as IR, they're likely going to bloat the LLVM module, which has a single-threaded execution model. Bloating the IR and trying to do whole-program analysis can be disastrous for compile-time performance.

I imagine the runtime will have very clear and simple entry points (create region, schedule behaviour, garbage collect, etc) and so LTO would have an easy job at inlining them later on with greater ease and no bloat in the middle-end.

Reply:
Updated, mentioning thinLTO, dropped the link time libraries